File: farp-ts/tests/path-match.test.ts
import { describe, it, expect } from 'vitest';
import {
  matchPath,
  shouldIncludePath,
  buildPathRules,
  INTERNAL_PATH_RULES,
  PathRuleAction,
} from '../src/index';
import type { PathRule } from '../src/index';

describe('path-match', () => {
  describe('matchPath', () => {
    it('matches exact paths', () => {
      expect(matchPath('/api/v1/users', '/api/v1/users')).toBe(true);
      expect(matchPath('/api/v1/users', '/api/v1/orders')).toBe(false);
    });

    it('matches single wildcard *', () => {
      expect(matchPath('/api/*/status', '/api/users/status')).toBe(true);
      expect(matchPath('/api/*/status', '/api/orders/status')).toBe(true);
      expect(matchPath('/api/*/status', '/api/users/health')).toBe(false);
    });

    it('single wildcard does not match multiple segments', () => {
      expect(matchPath('/api/*/status', '/api/v1/users/status')).toBe(false);
    });

    it('matches double wildcard ** at end', () => {
      expect(matchPath('/api/**', '/api')).toBe(true);
      expect(matchPath('/api/**', '/api/v1')).toBe(true);
      expect(matchPath('/api/**', '/api/v1/users/123')).toBe(true);
    });

    it('matches double wildcard ** in middle', () => {
      expect(matchPath('/api/**/status', '/api/status')).toBe(true);
      expect(matchPath('/api/**/status', '/api/v1/status')).toBe(true);
      expect(matchPath('/api/**/status', '/api/v1/v2/status')).toBe(true);
    });

    it('matches inline wildcards', () => {
      expect(matchPath('/openapi*', '/openapi.json')).toBe(true);
      expect(matchPath('/openapi*', '/openapi.yaml')).toBe(true);
      expect(matchPath('/openapi*', '/openapi')).toBe(true);
      expect(matchPath('/*json', '/openapi.json')).toBe(true);
    });

    it('matches root path', () => {
      expect(matchPath('/', '/')).toBe(true);
      expect(matchPath('/', '/api')).toBe(false);
    });

    it('handles trailing slashes', () => {
      expect(matchPath('/api/v1/', '/api/v1')).toBe(true);
    });
  });

  describe('shouldIncludePath', () => {
    it('returns true by default when no rules match', () => {
      const rules: PathRule[] = [
        { pattern: '/admin/**', action: PathRuleAction.Exclude },
      ];
      expect(shouldIncludePath('/api/users', rules)).toBe(true);
    });

    it('first matching rule wins - exclude', () => {
      const rules: PathRule[] = [
        { pattern: '/api/**', action: PathRuleAction.Exclude },
        { pattern: '/api/users', action: PathRuleAction.Include },
      ];
      expect(shouldIncludePath('/api/users', rules)).toBe(false);
    });

    it('first matching rule wins - include', () => {
      const rules: PathRule[] = [
        { pattern: '/api/users', action: PathRuleAction.Include },
        { pattern: '/api/**', action: PathRuleAction.Exclude },
      ];
      expect(shouldIncludePath('/api/users', rules)).toBe(true);
    });

    it('exclude rule prevents inclusion', () => {
      const rules: PathRule[] = [
        { pattern: '/internal/**', action: PathRuleAction.Exclude },
      ];
      expect(shouldIncludePath('/internal/metrics', rules)).toBe(false);
    });
  });

  describe('buildPathRules', () => {
    it('returns only user rules when excludeInternal is false', () => {
      const userRules: PathRule[] = [
        { pattern: '/custom/**', action: PathRuleAction.Exclude },
      ];
      const result = buildPathRules(false, userRules);
      expect(result).toEqual(userRules);
    });

    it('prepends internal rules when excludeInternal is true', () => {
      const userRules: PathRule[] = [
        { pattern: '/custom/**', action: PathRuleAction.Exclude },
      ];
      const result = buildPathRules(true, userRules);
      expect(result.length).toBe(INTERNAL_PATH_RULES.length + userRules.length);
      expect(result.slice(0, INTERNAL_PATH_RULES.length)).toEqual(INTERNAL_PATH_RULES);
      expect(result.slice(INTERNAL_PATH_RULES.length)).toEqual(userRules);
    });
  });

  describe('INTERNAL_PATH_RULES', () => {
    it('excludes root path', () => {
      expect(shouldIncludePath('/', INTERNAL_PATH_RULES)).toBe(false);
    });

    it('excludes _farp paths', () => {
      expect(shouldIncludePath('/_farp/manifest', INTERNAL_PATH_RULES)).toBe(false);
      expect(shouldIncludePath('/_farp/health', INTERNAL_PATH_RULES)).toBe(false);
    });

    it('excludes docs paths', () => {
      expect(shouldIncludePath('/docs', INTERNAL_PATH_RULES)).toBe(false);
      expect(shouldIncludePath('/docs/api', INTERNAL_PATH_RULES)).toBe(false);
    });

    it('excludes openapi and asyncapi files', () => {
      expect(shouldIncludePath('/openapi.json', INTERNAL_PATH_RULES)).toBe(false);
      expect(shouldIncludePath('/asyncapi.json', INTERNAL_PATH_RULES)).toBe(false);
    });

    it('includes normal API paths', () => {
      expect(shouldIncludePath('/api/users', INTERNAL_PATH_RULES)).toBe(true);
      expect(shouldIncludePath('/v1/orders', INTERNAL_PATH_RULES)).toBe(true);
    });
  });
});

File: path_match_test.go
package farp

import "testing"

func TestMatchPath(t *testing.T) {
	tests := []struct {
		pattern string
		path    string
		want    bool
	}{
		// Exact matches
		{"/health", "/health", true},
		{"/health", "/healthz", false},
		{"/api/v1/users", "/api/v1/users", true},
		{"/api/v1/users", "/api/v2/users", false},

		// Single-segment wildcard (*)
		{"/api/*/users", "/api/v1/users", true},
		{"/api/*/users", "/api/v2/users", true},
		{"/api/*/users", "/api/v1/posts", false},
		{"/api/*/users", "/api/users", false},       // * requires exactly one segment
		{"/api/*/users", "/api/v1/v2/users", false}, // * matches only one segment
		{"/api/users/*", "/api/users/123", true},
		{"/api/users/*", "/api/users/123/edit", false},

		// Multi-segment wildcard (**)
		{"/api/**", "/api", true}, // ** matches zero segments
		{"/api/**", "/api/v1", true},
		{"/api/**", "/api/v1/users", true},
		{"/api/**", "/api/v1/users/123", true},
		{"/internal/**", "/internal", true},
		{"/internal/**", "/internal/foo/bar/baz", true},
		{"/**", "/anything/at/all", true},
		{"/**", "/", true},

		// ** in the middle
		{"/api/**/status", "/api/status", true},
		{"/api/**/status", "/api/v1/status", true},
		{"/api/**/status", "/api/v1/users/status", true},
		{"/api/**/status", "/api/v1/users/health", false},

		// Mixed wildcards
		{"/api/*/users/**", "/api/v1/users", true},
		{"/api/*/users/**", "/api/v1/users/123", true},
		{"/api/*/users/**", "/api/v1/users/123/edit", true},
		{"/api/*/users/**", "/api/users/123", false}, // * requires one segment

		// Root path
		{"/", "/", true},
		{"/", "/api", false},

		// Pattern with trailing slash normalization
		{"/_farp/**", "/_farp/manifest", true},
		{"/_farp/**", "/_farp/schemas/openapi", true},
		{"/_/**", "/_/health", true},
		{"/_/**", "/_/info", true},

		// OpenAPI/AsyncAPI prefix patterns
		{"/openapi*", "/openapi.json", true},
		{"/openapi*", "/openapi.yaml", true},
		{"/asyncapi*", "/asyncapi.json", true},
		{"/docs/**", "/docs", true},
		{"/docs/**", "/docs/index.html", true},
		{"/docs", "/docs", true},
		{"/docs", "/docs/index.html", false},
	}

	for _, tt := range tests {
		t.Run(tt.pattern+"→"+tt.path, func(t *testing.T) {
			got := MatchPath(tt.pattern, tt.path)
			if got != tt.want {
				t.Errorf("MatchPath(%q, %q) = %v, want %v", tt.pattern, tt.path, got, tt.want)
			}
		})
	}
}

func TestShouldIncludePath(t *testing.T) {
	tests := []struct {
		name  string
		path  string
		rules []PathRule
		want  bool
	}{
		{
			name:  "no rules - include by default",
			path:  "/api/users",
			rules: nil,
			want:  true,
		},
		{
			name: "exclude rule matches",
			path: "/internal/debug",
			rules: []PathRule{
				{Pattern: "/internal/**", Action: PathRuleExclude},
			},
			want: false,
		},
		{
			name: "include rule matches",
			path: "/api/v1/users",
			rules: []PathRule{
				{Pattern: "/api/**", Action: PathRuleInclude},
			},
			want: true,
		},
		{
			name: "first match wins - exclude before include",
			path: "/api/internal/debug",
			rules: []PathRule{
				{Pattern: "/api/internal/**", Action: PathRuleExclude},
				{Pattern: "/api/**", Action: PathRuleInclude},
			},
			want: false,
		},
		{
			name: "first match wins - include before exclude",
			path: "/api/v1/users",
			rules: []PathRule{
				{Pattern: "/api/v1/**", Action: PathRuleInclude},
				{Pattern: "/api/**", Action: PathRuleExclude},
			},
			want: true,
		},
		{
			name: "no match - default include",
			path: "/other/path",
			rules: []PathRule{
				{Pattern: "/api/**", Action: PathRuleExclude},
			},
			want: true,
		},
		{
			name: "exact exclude",
			path: "/health",
			rules: []PathRule{
				{Pattern: "/health", Action: PathRuleExclude},
			},
			want: false,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := ShouldIncludePath(tt.path, tt.rules)
			if got != tt.want {
				t.Errorf("ShouldIncludePath(%q) = %v, want %v", tt.path, got, tt.want)
			}
		})
	}
}

func TestBuildPathRules(t *testing.T) {
	userRules := []PathRule{
		{Pattern: "/api/users/*", Action: PathRuleExclude},
	}

	// With internal exclusion enabled
	rules := BuildPathRules(true, userRules)
	if len(rules) <= len(userRules) {
		t.Error("expected internal rules to be prepended")
	}

	// Internal paths should be excluded
	if ShouldIncludePath("/_farp/manifest", rules) {
		t.Error("/_farp/manifest should be excluded with internal rules")
	}

	if ShouldIncludePath("/_/health", rules) {
		t.Error("/_/health should be excluded with internal rules")
	}

	// User rule should also work
	if ShouldIncludePath("/api/users/123", rules) {
		t.Error("/api/users/123 should be excluded by user rule")
	}

	// Non-matching paths should be included
	if !ShouldIncludePath("/api/v1/projects", rules) {
		t.Error("/api/v1/projects should be included")
	}

	// Without internal exclusion
	rulesNoInternal := BuildPathRules(false, userRules)
	if len(rulesNoInternal) != len(userRules) {
		t.Error("expected only user rules when internal exclusion disabled")
	}

	// Internal paths should be included when disabled
	if !ShouldIncludePath("/_farp/manifest", rulesNoInternal) {
		t.Error("/_farp/manifest should be included when internal exclusion disabled")
	}
}

func TestInternalPathRules(t *testing.T) {
	rules := InternalPathRules()

	excluded := []string{"/", "/_farp/manifest", "/_farp/schemas/openapi", "/_/health", "/_/info", "/docs", "/docs/index.html"}
	for _, path := range excluded {
		if ShouldIncludePath(path, rules) {
			t.Errorf("%q should be excluded by internal rules", path)
		}
	}

	included := []string{"/api/v1/users", "/api/v1/projects", "/graphql"}
	for _, path := range included {
		if !ShouldIncludePath(path, rules) {
			t.Errorf("%q should be included (not internal)", path)
		}
	}
}
